src/manager.rs

```diff
@@ -0,0 +1,532 @@
+        num: &str,
+        mut op1_fun: F1,
+        mut op2_fun: F2,
+    ) -> Result<Number, RedisError>
+    where
+        F1: FnMut(i64, i64) -> i64,
+        F2: FnMut(f64, f64) -> f64,
+    {
+        let in_value = &serde_json::from_str(num)?;
+        if let Value::Number(in_value) = in_value {
+            let mut res = None;
+            self.do_op(path, |v| {
+                let num_res = match (v.as_i64(), in_value.as_i64()) {
+                    (Some(num1), Some(num2)) => ((op1_fun)(num1, num2)).into(),
+                    _ => {
+                        let num1 = v.as_f64().unwrap();
+                        let num2 = in_value.as_f64().unwrap();
+                        Number::from_f64((op2_fun)(num1, num2)).unwrap()
+                    }
+                };
+                res = Some(Value::Number(num_res));
+                Ok(res.clone())
+            })?;
+            match res {
+                None => Err(RedisError::Str("path does not exists")),
+                Some(n) => match n {
+                    Value::Number(n) => Ok(n),
+                    _ => Err(RedisError::Str("return value is not a number")),
+                },
+            }
+        } else {
+            Err(RedisError::Str("bad input number"))
+        }
+    }
+
+    fn set_root(&mut self, v: Option<Value>) -> Result<(), RedisError> {
+        match v {
+            Some(inner) => {
+                let v = self.key.get_value::<RedisJSON>(&REDIS_JSON_TYPE)?;
+                match v {
+                    Some(v) => v.data = inner,
+                    None => self
+                        .key
+                        .set_value(&REDIS_JSON_TYPE, RedisJSON { data: inner })?,
+                }
+            }
+            None => {
+                self.key.delete()?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl WriteHolder<Value, Value> for KeyHolderWrite {
+    fn apply_changes(&mut self, ctx: &Context, command: &str) -> Result<(), RedisError> {
+        if ctx.notify_keyspace_event(NotifyEvent::MODULE, command, &self.key_name) != Status::Ok {
+            Err(RedisError::Str("failed notify key space event"))
+        } else {
+            ctx.replicate_verbatim();
+            Ok(())
+        }
+    }
+
+    fn delete(&mut self) -> Result<(), RedisError> {
+        self.key.delete()?;
+        Ok(())
+    }
+
+    fn get_value(&self) -> Result<Option<&mut Value>, RedisError> {
+        let key_value = self.key.get_value::<RedisJSON>(&REDIS_JSON_TYPE)?;
+        match key_value {
+            Some(v) => Ok(Some(&mut v.data)),
+            None => Ok(None),
+        }
+    }
+
+    fn set_value(&mut self, path: Vec<String>, mut v: Value) -> Result<bool, RedisError> {
+        let mut updated = false;
+        if path.is_empty() {
+            // update the root
+            self.set_root(Some(v))?;
+            updated = true;
+        } else {
+            self.update(&path, self.get_value().unwrap().unwrap(), |_v| {
```

Comment:
`KeyHolderWrite::get_value` (and `KeyHolderRead::get_value`) may return `Ok(None)` so double `unwrap()` may panic? (also used in more locations)

Reply:
Yes but if you reach here its not possible that you will get none and so panic is OK here (if its happened its a bug and we want to know about it)